scanpy/external/tl/_sam.py

```diff
@@ -0,0 +1,216 @@
+"""\
+Run the Self-Assembling Manifold algorithm
+"""
+from typing import Optional, Union
+
+from anndata import AnnData
+
+from ... import logging as logg
+
+
+def sam(
+    adata: AnnData,
+    max_iter: int = 10,
+    num_norm_avg: int = 50,
+    k: int = 20,
+    distance: str = 'correlation',
+    standardization: Optional[str] = 'Normalizer',
+    weight_pcs: bool = True,
+    npcs: Optional[int] = None,
+    n_genes: Optional[int] = None,
+    projection: Optional[str] = 'umap',
+    inplace: bool = True,
+    verbose: bool = True,
+) -> Optional[AnnData]:
+    """Self-Assembling Manifolds single-cell RNA sequencing analysis tool.
+
+    SAM iteratively rescales the input gene expression matrix to emphasize
+    genes that are spatially variable along the intrinsic manifold of the data.
+    It outputs the gene weights, nearest neighbor matrix, and a 2D projection.
+
+    The AnnData input should contain unstandardized, non-negative values.
+    Preferably, the data should be log-normalized and no genes should be filtered out.
+
+
+    Parameters
+    ----------
+
+    k - int, optional, default 20
+        The number of nearest neighbors to identify for each cell.
+
+    distance : string, optional, default 'correlation'
+        The distance metric to use when identifying nearest neighbors.
+        Can be any of the distance metrics supported by sklearn's 'pdist'.
+
+    max_iter - int, optional, default 10
+        The maximum number of iterations SAM will run.
+
+    projection - str, optional, default 'umap'
+        If 'tsne', generates a t-SNE embedding. If 'umap', generates a UMAP
+        embedding. Otherwise, no embedding will be generated.
+
+    standardization - str, optional, default 'Normalizer'
+        If 'Normalizer', use sklearn.preprocessing.Normalizer, which
+        normalizes expression data prior to PCA such that each cell has
+        unit L2 norm. If 'StandardScaler', use
+        sklearn.preprocessing.StandardScaler, which normalizes expression
+        data prior to PCA such that each gene has zero mean and unit
+        variance. Otherwise, do not normalize the expression data. We
+        recommend using 'StandardScaler' for large datasets with many
+        expected cell types and 'Normalizer' otherwise.
+
+    num_norm_avg - int, optional, default 50
+        The top 'num_norm_avg' dispersions are averaged to determine the
+        normalization factor when calculating the weights. This prevents
+        genes with large spatial dispersions from skewing the distribution
+        of weights.
+
+    weight_pcs - bool, optional, default True
+        If True, scale the principal components by their eigenvalues. In
+        datasets with many expected cell types, setting this to False might
+        improve the resolution as these cell types might be encoded by low-
+        variance principal components.
+
+    npcs - int, optional, default None,
```

Comment:
Please call this `n_pcs`. You use `n_genes` below and [NPCs](https://en.wikipedia.org/wiki/Non-player_character) are something else :wink: